### PR TITLE
mTLS client authentication support

### LINKED ISSUE #8508: missing documentation for HTTPS with 2way TLS (aka mutual TLS, aka client cert auth)
**Describe the bug**
I'm trying to use client certificate authentication with Quarkus, and I didn't find documentation for it. In fact, I didn't find documentation for 2 things:

1. How to enable client cert auth. Clearly the configuration properties are there ([trust store](https://quarkus.io/guides/all-config#quarkus-vertx-http_quarkus.http.ssl.certificate.trust-store-file), [client auth](https://quarkus.io/guides/all-config#quarkus-vertx-http_quarkus.http.ssl.client-auth)), but that's all. I'd expect this to be described in the [HTTP guide](https://quarkus.io/guides/http-reference).

    Most likely, the request for documentation from [the original mutual TLS pull request](https://github.com/quarkusio/quarkus/pull/4663#issuecomment-554355015) has never materialized.
2. How to authenticate users based on the client certificate (and assign them roles). I've briefly looked around in Quarkus and it seems this isn't even possible. Is that true? (I certainly hope not!)

    Please let me know if this is really missing, so I can create a feature request.

    In any case, I'd expect documentation for that in the [HTTP guide](https://quarkus.io/guides/http-reference) and also the [security guide](https://quarkus.io/guides/security#authenticating-via-http).

### PR DESCRIPTION
**Pull request by pedroigor**
Fixes #8508

### CODE REVIEW

**Comment by pedroigor** [comment]
@stuartwdouglas I'm not sure about your suggestion on creating a default identity provider bean. This one is not here.

But the other points we discussed in https://github.com/quarkusio/quarkus/issues/8508 are implemented.

I sent a PR to https://github.com/quarkusio/quarkus-security/pull/8 with the `CertificateCredential` and `CertificateAuthenticationRequest`.

**Comment by pedroigor** [comment]
@stuartwdouglas Added the `X509IdentityProvider` to `quarkus-security`.

The doc is failing to build due to a duplicate id. The reason is that I've copied the `client-auth` config option to `HttpBuildTimeConfig` so I can get the value at build time. I did not want to change anything else as this simple change works fine except for doc generation.

**Comment by stuartwdouglas** [comment]
You can't have a property as both build and runtime, that does not work, it would need to be removed from the runtime config.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/MtlsAuthenticationMechanism.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+public class MtlsAuthenticationMechanism implements HttpAuthenticationMechanism {
+
+    @Override
+    public Uni<SecurityIdentity> authenticate(RoutingContext context,
+            IdentityProviderManager identityProviderManager) {
+        HttpServerRequest request = context.request();
+
+        if (!request.isSSL()) {
+            // we don't handle insecure requests
+            return Uni.createFrom().failure(new AuthenticationCompletionException());
```
Comment (stuartwdouglas) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/MtlsAuthenticationMechanism.java:50]: This should just return an empty identity, it is still possible for the user to authenticate via another mechanism.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/MtlsAuthenticationMechanism.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        }
+
+        Certificate certificate;
+
+        try {
+            certificate = request.sslSession().getPeerCertificates()[0];
+        } catch (SSLPeerUnverifiedException e) {
+            return Uni.createFrom().optional(Optional.empty());
```
Comment (stuartwdouglas) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/MtlsAuthenticationMechanism.java:58]: Uni now has a better API for this (createFrom().empty() I think).

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]

File: bom/runtime/pom.xml [hunk 3]
```diff
@@ -183,7 +183,7 @@
         <mockito.version>3.3.3</mockito.version>
         <jna.version>5.3.1</jna.version>
         <antlr.version>4.7.2</antlr.version>
-        <quarkus-security.version>1.1.1.Final</quarkus-security.version>
+        <quarkus-security.version>1.1.2.Final-SNAPSHOT</quarkus-security.version>
```
Comment (stuartwdouglas) [inline bom/runtime/pom.xml:186]: I have released this.

**Comment by pedroigor** [comment]
@stuartwdouglas I've moved the property from runtime to build time config. The reason why I didn't do it before is that the hot reload logic manually create the runtime config when handling a failure during the initial start. So I did not want to touch that part given that just copying the config from runtime to build time seemed to work (except for the doc generation).

I tried to write a test for a hot reload test to check if the change to the config is OK. But I did not figure out how to test the scenario where the application fails the initial start.

**Comment by stuartwdouglas** [comment]
The failure to start is tested in the maven integration tests, io.quarkus.maven.it.DevMojoIT#testThatApplicationRecoversStartupIssue

**Review by stuartwdouglas — APPROVED** [review]

**Comment by stuartwdouglas** [comment]
lgtm but needs a rebase

**Comment by pedroigor** [comment]
@stuartwdouglas failures do not seem to be related:

```
[ERROR] Failures: 
[ERROR]   TestEndpointRunner.overrides:65->invokeNonOverload:81 This method should have been replaced
[INFO] 
[ERROR] Tests run: 7, Failures: 1, Errors: 0, Skipped: 0
```

Re-running jobs.

**Comment by Ladicek** [comment]
I don't see documentation changes in this PR. Is there another PR with documentation? Thanks!

**Comment by pedroigor** [comment]
@Ladicek Yeah, I was supposed to send it on Fri. Will send it today.

**Comment by pedroigor** [comment]
@Ladicek See https://github.com/quarkusio/quarkus/pull/10172. I also found this great blog https://quarkus.io/blog/quarkus-mutual-tls/.

**Comment by Ladicek** [comment]
Great, thanks!

**Pull request merged**